src/evolutionary_keras/optimizers.py

```diff
@@ -216,7 +217,158 @@ def run_step(self, x, y):
+    the keras environment. 
+
+    Parameters
+    ----------
+        `sigma_init`: int
+            Allows adjusting the initial sigma
+    """
+
+    def __init__(self, sigma_init=1, *args, **kwargs):
+        self.sigma_init = sigma_init
+        self.shape = None
+        self.length_flat_layer = None
+        self.weightscopy = None
+        self.trainable_weights_names = None
+        super(CMA, self).__init__(*args, **kwargs)
+
+    def get_shape(self):
+        # we do all this to keep track of the position of the trainable weights
+        self.trainable_weights_names = [
+            weights.name for weights in self.model.trainable_weights
+        ]
+
+        if self.trainable_weights_names == []:
+            raise TypeError("The model does not have any trainable weights!")
+
+        self.shape = []
+        self.weightscopy = deepcopy(self.model.weights)
+        remove_these_weights = []
+        for i in range(len(self.model.weights)):
+            weight = self.model.weights[i]
+            if weight.name not in self.trainable_weights_names:
+                remove_these_weights.append(i)
+        number_to_remove = len(remove_these_weights)
+        for i in range(number_to_remove):
+            del self.weightscopy[remove_these_weights[number_to_remove - 1 - i]]
+        self.shape = [weight.shape.as_list() for weight in self.weightscopy]
+        return self.shape
+
+    """ 
+    'weights_per_layer' creates 'self.lengt_flat_layer' which is a list conatining
+    the numer of weights in each layer of the network. 
+    """
+
+    def weights_per_layer(self):
+        """
+        The first values of 'self.length_flat_layer' is set to 0 which is helpful
+        in determining the range of weights in the function 'undo_flatten'. 
+        """
+        self.length_flat_layer = [0]
+        for layer in range(len(self.shape)):
+            flat_layer = np.reshape(self.weightscopy[layer].numpy(), [-1])
+            self.length_flat_layer.append(len(flat_layer))
+
+    """
+    'flatten' returns a 1 dimensional list of all weights in the keras model 
+    """
+
+    def flatten(self):
+        weights = self.weightscopy
+        flattened_weights = []
+        for layer in range(len(weights)):
+            flattened_weights.append(np.reshape(weights[layer].numpy(), [-1]))
+        flattened_weights = np.concatenate(flattened_weights)
+        return flattened_weights
+
+    """ 
+    'undo_flatten' does the inverse of 'flatten': it takes a 1 dimensional input
+    and returns a weight structure that can be loaded into the model. 
+    """
+
+    def undo_flatten(self, flattened_weights):
+        new_weights = []
+        for i in range(len(self.shape)):
+            flat_layer = flattened_weights[
+                self.length_flat_layer[i] : self.length_flat_layer[i]
+                + self.length_flat_layer[i + 1]
+            ]
+            new_weights.append(np.reshape(flat_layer, self.shape[i]))

```

Comment:
Same here, you can do `for i, layer_shape in enumerate(self.shape)` (I would call the array `self.shapes` since there are more than one).